drivers/gpio/gpio_sam4l.c

```diff
@@ -0,0 +1,305 @@
+	/* Setup Pull-down resistor. */
+	} else if (flags & GPIO_PULL_DOWN) {
+		/* Enable pull-down. */
+		gpio->PDERS = mask;
+	}
+
+	/* Enable the GPIO to control the pin (instead of a peripheral). */
+	gpio->GPERS = mask;
+
+	return 0;
+}
+
+static int gpio_sam_config(const struct device *dev,
+			   gpio_pin_t pin,
+			   gpio_flags_t flags)
+{
+	return gpio_sam_port_configure(dev, BIT(pin), flags);
+}
+
+static int gpio_sam_port_get_raw(const struct device *dev,
+				 uint32_t *value)
+{
+	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
+	Gpio * const gpio = cfg->regs;
+
+	*value = gpio->PVR;
+
+	return 0;
+}
+
+static int gpio_sam_port_set_masked_raw(const struct device *dev,
+					uint32_t mask,
+					uint32_t value)
+{
+	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
+	Gpio * const gpio = cfg->regs;
+
+	gpio->OVR = (gpio->PVR & ~mask) | (mask & value);
+
+	return 0;
+}
+
+static int gpio_sam_port_set_bits_raw(const struct device *dev,
+				      uint32_t mask)
+{
+	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
+	Gpio * const gpio = cfg->regs;
+
+	/* Set pins. */
+	gpio->OVRS = mask;
+
+	return 0;
+}
+
+static int gpio_sam_port_clear_bits_raw(const struct device *dev,
+					uint32_t mask)
+{
+	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
+	Gpio * const gpio = cfg->regs;
+
+	/* Clear pins. */
+	gpio->OVRC = mask;
+
+	return 0;
+}
+
+static int gpio_sam_port_toggle_bits(const struct device *dev,
+				     uint32_t mask)
+{
+	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
+	Gpio * const gpio = cfg->regs;
+
+	/* Toggle pins. */
+	gpio->OVRT = mask;
+
+	return 0;
+}
+
+static int gpio_sam_port_interrupt_configure(const struct device *dev,
+					     uint32_t mask,
+					     enum gpio_int_mode mode,
+					     enum gpio_int_trig trig)
+{
+	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
+	Gpio * const gpio = cfg->regs;
+
+	/* Disable the interrupt. */
+	gpio->IERC = mask;
+	/* Set default to both edges */
+	gpio->IMR0C = mask;
+	gpio->IMR1C = mask;
+
+	if (trig != GPIO_INT_TRIG_BOTH) {
+		if (trig == GPIO_INT_TRIG_HIGH) {
+			gpio->IMR0S = mask;
+			gpio->IMR1C = mask;
+		} else {
+			gpio->IMR0C = mask;
+			gpio->IMR1S = mask;
+		}
+	}
+
+	if (mode != GPIO_INT_MODE_DISABLED) {
+		/* Clear any pending interrupts */
+		gpio->IFRC = mask;
+		/* Enable the interrupt. */
+		gpio->IERS = mask;
+	}
+
+	return 0;
+}
+
+static int gpio_sam_pin_interrupt_configure(const struct device *dev,
+					    gpio_pin_t pin,
+					    enum gpio_int_mode mode,
+					    enum gpio_int_trig trig)
+{
+	return gpio_sam_port_interrupt_configure(dev, BIT(pin), mode, trig);
+}
+
+static void gpio_sam_isr(const struct device *dev)
+{
+	const struct gpio_sam_config * const cfg = DEV_CFG(dev);
+	Gpio * const gpio = cfg->regs;
+	struct gpio_sam_runtime *context = dev->data;
+	uint32_t int_stat;
+
+	int_stat = gpio->IFR;
+
+	gpio_fire_callbacks(&context->cb, dev, int_stat);
+
+	gpio->IFRC = int_stat;
```

Comment:
The IFR needs to be cleared immediately after reading its value, otherwise we run the risk of losing an interrupt. We should assume that GPIO callbacks will be executing relatively long.

Reply:
Agreed, fixed!